Design System: Show context menu in a portal

samwhale opened the pull request:
## Context

~The context menu rendered an invisible mask that covered the entire screen (behind the menu). When a user clicked on this mask, the menu closed. However, to interact with the app they would have to click the app again. This wasn't great UX.~

Turns out this is how browsers usually do it. We moved the menu to be in a portal so that clicking anywhere on the screen would close it.

## Summary

1. ~Context menu now calls `onDismiss` when it loses focus. The `Mask` component has been removed and users are able to interact with the app even if the context menu is open in the dashboard and editor.~ The context menu is now displayed through a portal. Clicking anywhere on the screen will close it. Right clicking the mask will also close it.
2. Combined the `AnimatedContextMenu` and `ContextMenu` into one component. The animation can be used by using the `animate` prop.

## Relevant Technical Choices

n/a

## To-do

n/a

## User-facing changes



## Testing Instructions

New Functionality:
- Clicking away from the menu should close it but also interact with the thing you're clicking on. (ex: button or another element)

Everything else should work the same.

Dashboard:
- Should be able to open the menu by clicking or with the keyboard in the `My Stories` view
- Should be able to navigate the menu with keyboard.
- Tabbing away from the menu should close it.
- Arrow keys should not close the menu.
- When navigating with keyboard you should see the focus border.
- All actions in the menu should work as before.

Editor:
- Should be able to open the context menu by 'right-clicking' anywhere in the canvas. This should select the element that is under the right click.
- Right clicking another element should close the previous menu and open a new menu. The clicked element should now be selected
- All actions in the menu should work as before.

### QA

<!-- ignore-task-list-start -->
- [ ] This is a non-user-facing change and requires no QA
<!-- ignore-task-list-end -->

QA should only happen in the dashboard. The right-click menu in the editor will be QA'd all at once

## Reviews

### Does this PR have a security-related impact?

no

### Does this PR change what data or activity we track or use?

no

### Does this PR have a legal-related impact?

no

## Checklist

<!-- Check these after PR creation -->

- [x] This PR addresses an existing issue and I have linked this PR to it in ZenHub
- [x] I have tested this code to the best of my abilities
- [x] I have verified accessibility to the best of my abilities ([docs](https://github.com/google/web-stories-wp/blob/main/docs/accessibility-testing.md))
- [x] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [x] This PR contains automated tests (unit, integration, and/or e2e) to verify the code works as intended ([docs](https://github.com/google/web-stories-wp/tree/main/docs#testing))
- [x] I have added documentation where necessary
- [x] I have added a matching `Type: XYZ` label to the PR

---

<!--
Please reference the issue(s) this PR addresses.
No URLs, just the issue numbers.
Use "Fixes #123" if it fixes an issue.
Example:

Fixes #123
Partially addresses #456
See #789
-->

Fixes #8702

samwhale committed:
Combine AnimatedContextMenu and ContextMenu

samwhale committed:
Remove mask, add blurring to close

samwhale commented on packages/design-system/src/components/contextMenu/contextMenu.js:
```diff
@@ -17,29 +17,33 @@
+
+const ContextMenu = ({ animate, isAlwaysVisible, items, ...props }) => {
+  const Wrapper = useMemo(
+    () => (animate ? AnimationContainer : Popover),
```
Animation is now determined by prop rather than needing to import the `AnimatedContextMenu`

samwhale commented on packages/design-system/src/components/contextMenu/menu.js:
```diff
@@ -277,7 +303,13 @@ const Menu = ({
+      isIconMenu={isIconMenu}
+      role="menu"
+      onBlur={handleBlur}
+      tabIndex={0}
```
This is the magic here :sparkles:

samwhale committed:
Fix test

samwhale commented on packages/design-system/src/components/contextMenu/menu.js:
```diff
@@ -277,7 +303,13 @@ const Menu = ({
+      ref={wrapperRef}
+      isIconMenu={isIconMenu}
+      role="menu"
+      onBlur={handleBlur}
```
Looks like `onBlur` now hooks into `focusout` in react 17+: https://blog.saeloun.com/2021/05/14/react-17-uses-browse-focusin-focusout-for-onfocus-onblur

This allows us to detect blur events from children as well as the parent, since `focusout` bubbles and `blur` does not.

BrittanyIRL reviewed (commented):
Dahboard works great! that old issue with the keyboard didn't resurface which is awesome. 

🤔  With the right click menu, it feels a little weird that the right click menu pretty much never closes once it's open. For instance, I right click on a text element and then i decide to add a new image. The right click menu does swap out from being for text to an image element but I would expect the menu to close when i click on the image in the media panel and then if i want to right click the image i'd do so.  Could the menu just close when blur occurs not update to a new type? Is there a specific case that I'm missing?

samwhale committed:
Add the mask back in, remove blurring

samwhale committed:
Fix the menu finally
- close the menu when clicking anywher eon mask
- add menu to portal
- right clicking mask also closes menu

samwhale committed:
Fix karma tests

samwhale committed:
Fix js tests and add new test

samwhale commented on packages/design-system/src/components/contextMenu/mask.js:
```diff
@@ -36,13 +37,30 @@ const ScreenMask = styled.div`
 `;
 
 export default function Mask({ onDismiss }) {
+  const maskRef = useRef();
+
+  // Right clicking the mask should also close it
+  useEffect(() => {
+    const node = maskRef.current;
+    if (!node) {
+      return undefined;
+    }
+
+    node.addEventListener('contextmenu', onDismiss);
+
+    return () => {
+      node.removeEventListener('contextmenu', onDismiss);
+    };
+  }, [onDismiss]);
```
Right clicking the mask will calll `onDismiss` now too

samwhale commented on packages/story-editor/src/components/canvas/karma/rightClickMenu.karma.js:
```diff
@@ -46,6 +51,118 @@ describe('Right Click Menu integration', () => {
     fixture.restore();
   });
 
+  function rightClickMenu() {
+    return fixture.screen.getByRole('group', {
+      name: 'Context Menu for the selected element',
+    });
+  }
+
+  function sendBackward() {
+    return fixture.screen.getByRole('button', {
+      name: /^Send Backward/i,
+    });
+  }
+
+  function sendToBack() {
+    return fixture.screen.getByRole('button', {
+      name: /^Send to Back/i,
+    });
+  }
```
The right click menu is in a portal, so these were moved to search on the screen so that the tests would pass.

samwhale committed:
Fix test

BrittanyIRL commented on packages/design-system/src/components/contextMenu/menu.js:
```diff
@@ -181,12 +181,24 @@ const Menu = ({
   const ids = useMemo(() => items.map(() => uuidv4()), [items]);
 
-  const totalIndex = useMemo(() => items.length - 1, [items]);
+  // Need ref so that `items.length` changes do not trigger the effect
```
this is a good note 💯

BrittanyIRL reviewed (approved):
This turned out so great!  The discovery you did earlier last week and this ending solution is lovely. Thank you for updating tests and for the specific QA directions. Tested locally and dashboard has no regression and the editor's new right click menus all work well. 

I think for QA, let's specify regression testing for the menu on the dashboard. The right click menu will get tested all together with the flag ticket.
